src/host/posix/dnssd.hpp

```diff
@@ -112,25 +128,138 @@ class DnssdPlatform : public Mdns::StateObserver, private NonCopyable
     void  UnregisterHost(const Host &aHost, RequestId aRequestId, RegisterCallback aCallback);
     void  RegisterKey(const Key &aKey, RequestId aRequestId, RegisterCallback aCallback);
     void  UnregisterKey(const Key &aKey, RequestId aRequestId, RegisterCallback aCallback);
+    void  StartServiceBrowser(const Browser &aBrowser, const BrowseCallback &aCallback);
+    void  StopServiceBrowser(const Browser &aBrowser);
+    void  StartServiceResolver(const SrvResolver &aSrvResolver, const SrvCallback &aCallback);
+    void  StopServiceResolver(const SrvResolver &aSrvResolver);
+    void  StartTxtResolver(const TxtResolver &aTxtResolver, const TxtCallback &aCallback);
+    void  StopTxtResolver(const TxtResolver &aTxtResolver);
+    void  StartIp6AddressResolver(const AddressResolver &aAddressResolver, const AddressCallback &aCallback);
+    void  StopIp6AddressResolver(const AddressResolver &aAddressResolver);
+    void  StartIp4AddressResolver(const AddressResolver &aAddressResolver, const AddressCallback &aCallback);
+    void  StopIp4AddressResolver(const AddressResolver &aAddressResolver);
 
 private:
     static constexpr State kStateReady   = OT_PLAT_DNSSD_READY;
     static constexpr State kStateStopped = OT_PLAT_DNSSD_STOPPED;
 
+    class DnsName
+    {
+    public:
+        DnsName(std::string aName)
+            : mName(std::move(aName))
+        {
+        }
+
+        bool operator==(const DnsName &aOther) const
+        {
+            return StringUtils::ToLowercase(mName) == StringUtils::ToLowercase(aOther.mName);
+        }
+
+        bool operator<(const DnsName &aOther) const
+        {
+            return StringUtils::ToLowercase(mName) < StringUtils::ToLowercase(aOther.mName);
+        }
+
+    private:
+        std::string mName;
+    };
+
+    class DnsServiceName
```

Comment:
Similar to `DnsServiceName`, perhaps we could define a new class (e.g., `ServiceType`) to represent a service type with an optional sub-type. This class would basically contain two `DnsName` objects, where the sub-type can be empty or null.

We could then use this new `ServiceType` class as the key in `mServiceBrowsersTable`, mapping it to the `callbacktable`.

This should also allow us to handle sub-type Browse, though we will probably need to construct the full sub-type name from its pieces before passing it to the `ot-br` `mPublisher`. I assume the `mPublisher` requires the format `"<sub-type-label>._sub.<service_type>"` and can then do sub-type browse?